src/tools/rbd_mirror/GroupReplayer.cc

```diff
@@ -1232,6 +1260,175 @@ void GroupReplayer<I>::update_mirror_group_replay_status() {
   }
 }
 
+template <typename I>
+void GroupReplayer<I>::update_mirror_group_status(
+    bool force, const OptionalState &opt_state) {
+  dout(15) << "force=" << force << ", "
+           << "state=" << opt_state << dendl;
+
+  {
+    std::lock_guard locker{m_lock};
+    if (!force && !is_stopped_() && !is_running_()) {
+      dout(15) << "shut down in-progress: ignoring update" << dendl;
+      return;
+    }
+  }
+
+  m_in_flight_op_tracker.start_op();
+  auto ctx = new LambdaContext(
+    [this, force, opt_state](int r) {
+      set_mirror_group_status_update(force, opt_state);
+      m_in_flight_op_tracker.finish_op();
+    });
+  m_threads->work_queue->queue(ctx, 0);
+}
+
+template <typename I>
+void GroupReplayer<I>::set_mirror_group_status_update(
+  bool force, const OptionalState &opt_state) {
+  dout(15) << "force=" << force << ", "
+           << "state=" << opt_state << dendl;
+
+  reregister_admin_socket_hook();
+
+  State state;
+  std::string state_desc;
+  int last_r;
+  bool stopping_replay;
+  bool skip_image_status_update = false;
+
+  auto mirror_group_status_state = boost::make_optional(
```

Comment:
This variable is eventually used to determine whether the replayer is in one of the following states,
 group_replayer::HEALTH_STATE_OK,  group_replayer::HEALTH_STATE_WARNING, group_replayer::HEALTH_STATE_ERROR  in `get_health_state()`
 
 I wish this was just an enum of type `group_replayer::HEALTH_STATE`, with default value of  "HEALTH_STATE_OK" with equivalent class data member `m_group_health_state` let's say.
 
 Then set this variable to
 "HEALTH_STATE_ERROR" within the condition `last_r < 0 && last_r != -ECANCELED`
 and "HEALTH_STATE_WARNIN" within the condition `last_r == -EREMOTEIO` .
 
 We don't need to do this now. Just noting my thoughts. What do you think?

Reply:
The current approach can definitely be confusing but I wanted to make as few changes as possible at this time.  In future we can look at doing this differently in both ImageReplayer(which uses the same approach) and GroupReplayer.

Reply:
@idryomov what do you think about my suggestion? I can take this up this as a separate PR

Reply:
Would the new `m_image_health_state` replace the existing `m_mirror_image_status_state` class member?  If not, 
I'm not sure it's worth it because we would end up with two class members tracking more or less the same thing. The (possibly superficial) benefit of the current approach is that the mapping from image status to health state is fully contained within `get_health_state()`.

If yes, I'm all for it -- it would also automatically get rid of that confusing use of `boost::make_optional()`.

Reply:
@idryomov  ah okay! I think I can replace `mirror_group_status_state` with `m_group_health_state` . I'll work on this